### PR TITLE
Use pallet-contracts-uapi

### PR DESCRIPTION
**Pull request by pgherveou**
We now have a `pallet-contracts-uapi` crate (see issue https://github.com/paritytech/polkadot-sdk/issues/2189)  that expose all the host functions that a contract can import.

This PR remove the functions that were originally defined in [https://github.com/paritytech/ink/blob/master/crates/env/src/engine/on_chain/ext.rs](ext.rs) and use the one defined in this crate instead.

Some shared structs like Return and CallFlags are also now imported from this crate instead of being redefined in ink!

Note: Depends on the next crates.io release of `pallet-contracts-uapi`

### CODE REVIEW

File: crates/env/src/engine/mod.rs [hunk 1]
```diff
@@ -193,7 +195,11 @@ mod decode_instantiate_result_tests {
             DefaultEnvironment,
             TestContractRef,
             Result<TestContractRef, ContractError>,
-        >(Err(Error::CalleeReverted), out_address, out_return_value)
+        >(
+            Err(Error::ReturnError(ReturnErrorCode::CalleeReverted)),
```
Comment (ascjones) [inline crates/env/src/engine/mod.rs:199]: And here

File: crates/env/src/lib.rs [hunk 2]
```diff
@@ -102,16 +102,18 @@ mod tests;
 #[doc(inline)]
 pub use self::engine::off_chain::test_api as test;
 
+pub use pallet_contracts_uapi::{
```
Comment (ascjones) [inline crates/env/src/lib.rs:105]: How will the docs look, should we do `#[doc(inline)]` for these types?

**Review by ascjones — COMMENTED** [review]
All in alllooks pretty good. 

It would still be interesting to know in a couple of cases (e.g. `call-builder-return-value` and `ontract-storage`) why the contracts are 250 bytes or so bigger. I remember you looking into it previosly and overall the diffs don't seem as big now, and in many cases small reductions in size.

File: crates/env/src/lib.rs (hunk 2 above)
Comment (pgherveou) [reply crates/env/src/lib.rs:105]: ah did not know we could do that, let me  test to see how the gen docs looks like

File: crates/env/src/engine/mod.rs [hunk 3]
```diff
@@ -89,7 +91,7 @@ where
             let output = <R as ConstructorReturnType<ContractRef>>::ok(contract_ref);
             Ok(Ok(output))
         }
-        Err(EnvError::CalleeReverted) => {
+        Err(EnvError::ReturnError(ReturnErrorCode::CalleeReverted)) => {
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply crates/env/src/engine/mod.rs:94]: you can't do that in a pattern matching can you?

**Review by ascjones — APPROVED** [review]
LGTM

**Comment by pgherveou** [comment]
mm pallet-contracts-uapi@5.0.0 is actually a buggy version, thus the tests are not passing here we will have to wait for the next release of the crate

**Comment by pgherveou** [comment]
@ascjones, I published the package under `ink-pallet-contracts-uapi` so we can merge it now if that's ok for you

**Pull request merged**